zebra-chain/src/work/equihash.rs

```diff
@@ -60,11 +65,120 @@ impl Solution {
+    /// The `solution` in the header template is ignored.
+    ///
+    /// This method is CPU and memory-intensive. It uses 144 MB of RAM and one CPU core while running.
+    /// It can run for minutes or hours if the network difficulty is high.
+    #[cfg(feature = "internal-miner")]
+    #[allow(clippy::unwrap_in_result)]
+    pub fn solve<F>(mut header: Header, mut cancel_fn: F) -> Result<Header, SolverCancelled>
+    where
+        F: FnMut() -> Result<(), SolverCancelled>,
+    {
+        use crate::shutdown::is_shutting_down;
+
+        let mut input = Vec::new();
+        header
+            .zcash_serialize(&mut input)
+            .expect("serialization into a vec can't fail");
+        let input = &input[0..Solution::INPUT_LENGTH];
+
+        while !is_shutting_down() {
+            // Don't run the solver if we'd just cancel it anyway.
+            cancel_fn()?;
+
+            let solutions = equihash_solver::tromp::solve_200_9_compressed(input, || {
+                // Cancel the solver if we have a new template.
+                if cancel_fn().is_err() {
+                    return None;
+                }
+
+                // This skips the first nonce, which doesn't matter in practice.
+                Self::next_nonce(&mut header.nonce);
+                Some(*header.nonce)
+            });
+
+            // If we got any solutions, try submitting them, because the new template might just
+            // contain some extra transactions. Mining extra transactions is optional.
+            for solution in &solutions {
+                header.solution = Self::from_bytes(solution)
+                    .expect("unexpected invalid solution: incorrect length");
+
+                // TODO: work out why we sometimes get invalid solutions here
+                if let Err(error) = header.solution.check(&header) {
+                    info!(?error, "found invalid solution for header");
+                    continue;
+                }
+
+                if Self::difficulty_is_valid(&header) {
+                    info!("found valid solution and difficulty");
+                    return Ok(header);
+                }
+            }
+
+            debug!(
+                solutions = ?solutions.len(),
+                "found valid solutions which did not pass the validity or difficulty checks"
+            );
+        }
+
+        Err(SolverCancelled)
+    }
+
+    /// Modifies `nonce` to be the next integer in big-endian order.
+    /// Wraps to zero if the next nonce would overflow.
+    #[cfg(feature = "internal-miner")]
+    fn next_nonce(nonce: &mut [u8; 32]) {
+        let _ignore_overflow = crate::primitives::byte_array::increment_big_endian(&mut nonce[..]);
+    }
+
+    /// Returns `true` if the `nonce` and `solution` in `header` meet the difficulty threshold.
+    ///
+    /// Assumes that the difficulty threshold in the header is valid.
+    #[cfg(feature = "internal-miner")]
+    fn difficulty_is_valid(header: &Header) -> bool {
```

Comment:
Optional/Nitpick: Could we use the same method for this and the block verifier check here: https://github.com/ZcashFoundation/zebra/blob/internal-mine/zebra-consensus/src/block/check.rs#L100

Reply:
No, because the original method checks the difficult threshold as well, which is unnecessary here. (We assume that the caller specified the threshold correctly in the header template, and if they didn't, the block will get rejected during full validation.)

Also, the error type returned by the original method is only in the state crate.

We could extract out this check into a separate method, but trying to get the APIs to match could be tricky. So it seemed easiest just to copy 2 lines of code and say where they came from.

The reference you asked for is already in a comment on line 169 of this file.